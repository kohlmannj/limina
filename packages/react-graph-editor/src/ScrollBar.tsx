import React, { CSSProperties, SFC } from 'react';
import styled, { css, StyledComponent } from 'react-emotion';
import { OverflowMode, ScrollBarOrientation } from './index';

export interface IScrollBarProps {
  className?: string;
  orientation?: ScrollBarOrientation;
  overflow?: OverflowMode;
  progress?: number;
  scale?: number;
  style?: CSSProperties;
}

export interface IScrollBarTheme {
  thumbColor?: string;
  thumbWidth?: number;
  trackColor?: string;
  trackWidth?: number;
}

const getOvershootFactor = (progress = 0, min = 0, max = 1): number => {
  let overshoot;

  if (progress < min) {
    overshoot = 1 - Math.abs(progress);
  } else if (progress >= max) {
    overshoot = 1 - Math.abs(progress - max);
  } else {
    overshoot = 1;
  }

  return overshoot;
};

const positionForOrientation = (props: IStyledScrollBarProps) => css`
  left: ${!props.orientation || props.orientation === 'vertical' ? 'auto' : 0};
  right: ${!props.orientation || props.orientation === 'vertical'
    ? 0
    : `${
        props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
      }px`};
  top: ${!props.orientation || props.orientation === 'vertical' ? 0 : 'auto'};
  bottom: ${!props.orientation || props.orientation === 'vertical'
    ? `${props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`
    : 0};
`;

const Track = styled<IScrollBarProps, 'div'>('div')`
  background: ${props => props.theme.trackColor || `#efefef`};
  display: flex;
  flex-direction: ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'column' : 'row'};
  align-items: center;
  justify-content: ${props =>
    typeof props.progress === 'number' && props.progress >= 0.5 ? 'flex-end' : 'flex-start'};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`};
  ${positionForOrientation};
`;

const ThumbContainer = styled<IScrollBarProps, 'div'>('div')`
  display: flex;
  flex-direction: inherit;
  justify-content: inherit;
  align-items: center;
  --overshoot-factor: ${props => getOvershootFactor(props.progress)};
  ${props => (!props.orientation || props.orientation === 'vertical' ? 'width' : 'height')}: 100%;
  /* Thumb Length */
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'height' : 'width'}: ${props =>
    `calc(1 / ${props.scale || 1} * ${getOvershootFactor(props.progress)} * 100%)`};
  transform: ${props =>
    `translate${!props.orientation || props.orientation === 'vertical' ? 'Y' : 'X'}(${(Math.min(
      Math.max(props.progress || 0, 0),
      1
    ) +
      (typeof props.progress === 'number' && props.progress >= 0.5 ? -1 : 0)) *
    ((props.scale || 1) -
      1) /* *
      (typeof props.progress === 'number' && props.progress > 1
        ? (props.progress || 1) / getOvershootFactor(props.progress)
        : 1) */ *
      100}%)`};
`;

export interface IStyledScrollBarProps extends IScrollBarProps {
  theme?: IScrollBarTheme;
}

export const Thumb: StyledComponent<IScrollBarProps, IScrollBarTheme, {}> = styled<
  IScrollBarProps,
  'button'
>('button')`
  background: ${props => props.theme.thumbColor || '#666'};
  border-radius: ${props =>
    `${(typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15) / 2}px`};
  display: flex;
  padding: 0;
  border: 0;
  flex-direction: inherit;
  flex: 0 1 100%;
  margin: ${props =>
    `${!props.orientation || props.orientation === 'vertical' ? '' : '0 '}${((typeof props.theme
      .trackWidth === 'number'
      ? props.theme.trackWidth
      : 15) -
      (typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15)) /
      2}px${!props.orientation || props.orientation === 'vertical' ? ' 0' : ''}`};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15}px`};
`;

export const ScrollBar: SFC<IScrollBarProps> = ({
  className,
  orientation,
  overflow,
  progress,
  scale,
  style,
}) => (
  <Track className={className} orientation={orientation} progress={progress} style={style}>
    <ThumbContainer orientation={orientation} progress={progress} scale={scale}>
      {typeof scale === 'number' &&
        (scale > 1 || (scale <= 1 && overflow === 'scroll')) && <Thumb orientation={orientation} />}
    </ThumbContainer>
  </Track>
);

ScrollBar.defaultProps = {
  orientation: 'vertical',
  overflow: 'auto',
  scale: 1,
};

export default ScrollBar;
