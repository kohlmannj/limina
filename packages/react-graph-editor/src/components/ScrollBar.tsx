import React, { CSSProperties, SFC } from 'react';
import styled, { css, StyledComponent } from 'react-emotion';
import { OverflowMode, ScrollBarOrientation } from '..';

export interface IScrollBarProps {
  className?: string;
  orientation?: ScrollBarOrientation;
  overflow?: OverflowMode;
  progress?: number;
  scale?: number;
  style?: CSSProperties;
}

export interface IScrollBarTheme {
  thumbColor?: string;
  thumbWidth?: number;
  trackColor?: string;
  trackWidth?: number;
}

const getOvershootFactor = (progress = 0, min = 0, max = 1): number => {
  let overshoot;

  if (progress < min) {
    overshoot = 1 - Math.abs(progress);
  } else if (progress >= max) {
    overshoot = 1 - Math.abs(progress - max);
  } else {
    overshoot = 1;
  }

  return overshoot;
};

const positionForOrientation = (props: IStyledScrollBarProps) => css`
  top: ${!props.orientation || props.orientation === 'vertical' ? 0 : 'auto'};
  right: ${!props.orientation || props.orientation === 'vertical'
    ? 0
    : `${
        props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
      }px`};
  bottom: ${!props.orientation || props.orientation === 'vertical'
    ? `${props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`
    : 0};
  left: ${!props.orientation || props.orientation === 'vertical' ? 'auto' : 0};
`;

const Track = styled<IScrollBarProps, 'div'>('div')`
  display: flex;
  flex-direction: ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'column' : 'row'};
  align-items: center;
  justify-content: ${props =>
    typeof props.progress === 'number' && props.progress >= 0.5 ? 'flex-end' : 'flex-start'};
  background: ${props => props.theme.trackColor || `#efefef`};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`};
  ${positionForOrientation};
`;

const ThumbContainer = styled<IScrollBarProps, 'div'>('div')`
  display: flex;
  flex-direction: inherit;
  align-items: center;
  justify-content: inherit;
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width: 100%' : 'height: 100%'};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'height' : 'width'}: ${props =>
    `calc(1 / ${props.scale || 1} * ${getOvershootFactor(props.progress)} * 100%)`};
  transform: ${props =>
    `translate${!props.orientation || props.orientation === 'vertical' ? 'Y' : 'X'}(${(Math.min(
      Math.max(props.progress || 0, 0),
      1
    ) +
      (typeof props.progress === 'number' && props.progress >= 0.5 ? -1 : 0)) *
      ((props.scale || 1) - 1) *
      100}%)`};
`;

export interface IStyledScrollBarProps extends IScrollBarProps {
  theme?: IScrollBarTheme;
}

export const Thumb = styled<IScrollBarProps, 'button'>('button')`
  display: flex;
  flex: 0 1 100%;
  flex-direction: inherit;
  padding: 0;
  border: 0;
  margin: ${props =>
    `${!props.orientation || props.orientation === 'vertical' ? '' : '0 '}${((typeof props.theme
      .trackWidth === 'number'
      ? props.theme.trackWidth
      : 15) -
      (typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15)) /
      2}px${!props.orientation || props.orientation === 'vertical' ? ' 0' : ''}`};
  background: ${props => props.theme.thumbColor || '#666'};
  border-radius: ${props =>
    `${(typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15) / 2}px`};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15}px`};
`;

export const ScrollBar: SFC<IScrollBarProps> = ({
  className,
  orientation,
  overflow,
  progress,
  scale,
  style,
}) => (
  <Track className={className} orientation={orientation} progress={progress} style={style}>
    <ThumbContainer orientation={orientation} progress={progress} scale={scale}>
      {typeof scale === 'number' &&
        (scale > 1 || (scale <= 1 && overflow === 'scroll')) && <Thumb orientation={orientation} />}
    </ThumbContainer>
  </Track>
);

ScrollBar.defaultProps = {
  orientation: 'vertical',
  overflow: 'auto',
  scale: 1,
};

export default ScrollBar;
