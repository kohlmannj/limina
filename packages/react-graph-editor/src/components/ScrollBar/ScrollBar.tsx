import React, { CSSProperties, SFC } from 'react';
import { OverflowMode, ScrollBarAxis } from '../..';
import { IScrollBarTheme } from '../../theme';
import ThumbContainer from './components/ThumbContainer';
import ThumbSegment, { IStyledThumbSegmentProps } from './components/ThumbSegment';
import Track from './components/Track';

export interface IScrollBarProps {
  axis?: ScrollBarAxis;
  className?: string;
  overflow?: OverflowMode;
  progress?: number;
  scale?: number;
  style?: CSSProperties;
  thumbEndProps?: Partial<IStyledThumbSegmentProps>;
  thumbMiddleProps?: Partial<IStyledThumbSegmentProps>;
  thumbStartProps?: Partial<IStyledThumbSegmentProps>;
}

export interface IStyledScrollBarProps extends IScrollBarProps {
  theme?: Partial<IScrollBarTheme>;
}

export const ScrollBar: SFC<IScrollBarProps> = ({
  axis,
  className,
  overflow,
  progress,
  scale,
  style,
  thumbEndProps,
  thumbMiddleProps,
  thumbStartProps,
}) => (
  <Track className={className} axis={axis} progress={progress} style={style}>
    {typeof scale !== 'undefined' &&
      (scale > 1 || (scale <= 1 && overflow === 'scroll')) && (
        <ThumbContainer axis={axis} progress={progress} scale={scale}>
          <ThumbSegment dragSignifier axis={axis} position="start" {...thumbStartProps} />
          <ThumbSegment axis={axis} position="middle" {...thumbMiddleProps} />
          <ThumbSegment dragSignifier axis={axis} position="end" {...thumbEndProps} />
        </ThumbContainer>
      )}
  </Track>
);

ScrollBar.defaultProps = {
  axis: 'y',
  overflow: 'scroll',
  scale: 1,
};

export default ScrollBar;
