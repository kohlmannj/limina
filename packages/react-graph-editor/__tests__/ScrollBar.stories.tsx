/* tslint:disable no-implicit-dependencies */
import { color, number, select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'emotion-theming';
import React from 'react';
import ScrollBar from '../src/ScrollBar';

storiesOf('react-graph-editor/ScrollBar', module)
  .add('with knobs', () => {
    const orientation = select(
      'Orientation',
      { vertical: 'Vertical', horizontal: 'Horizontal' },
      'vertical',
      'orientation'
    );

    return (
      <ThemeProvider
        theme={{
          thumbColor: color('Thumb Color', '#666', 'thumbColor'),
          trackColor: color('Track Color', '#efefef', 'trackColor'),
          thumbWidth: number(
            'Thumb Width',
            15,
            {
              range: true,
              min: 0,
              max: 100,
              step: 0.5,
            },
            'thumbWidth'
          ),
          trackWidth: number(
            'Track Width',
            15,
            {
              range: true,
              min: 0,
              max: 100,
              step: 1,
            },
            'trackWidth'
          ),
        }}
      >
        <ScrollBar
          progress={number(
            'Progress',
            0,
            {
              range: true,
              min: -0.5,
              max: 1.5,
              step: 0.01,
            },
            'progress'
          )}
          scale={number(
            'Scale',
            2,
            {
              range: true,
              min: 1,
              max: 10,
              step: 1,
            },
            'scale'
          )}
          orientation={orientation}
          overflow={select('Overflow', { auto: 'Auto', scroll: 'Scroll' }, 'auto', 'overflow')}
          style={{ [orientation === 'vertical' ? 'height' : 'width']: '400px', margin: '25vh' }}
        />
      </ThemeProvider>
    );
  })
  .add('with default props', () => <ScrollBar style={{ height: '100vh' }} />);

storiesOf('react-graph-editor/ScrollBar/Vertical/Scale 2', module)
  .add('with progress 0', () => <ScrollBar scale={2} style={{ height: '100vh' }} />)
  .add('with progress 0.25', () => (
    <ScrollBar progress={0.25} scale={2} style={{ height: '100vh' }} />
  ))
  .add('with progress 0.5', () => (
    <ScrollBar progress={0.5} scale={2} style={{ height: '100vh' }} />
  ))
  .add('with progress 0.75', () => (
    <ScrollBar progress={0.75} scale={2} style={{ height: '100vh' }} />
  ))
  .add('with progress 1', () => <ScrollBar progress={1} scale={2} style={{ height: '100vh' }} />);

storiesOf('react-graph-editor/ScrollBar/Vertical/Scale 10', module)
  .add('with progress 0', () => <ScrollBar scale={10} style={{ height: '100vh' }} />)
  .add('with progress 0.25', () => (
    <ScrollBar progress={0.25} scale={10} style={{ height: '100vh' }} />
  ))
  .add('with progress 0.5', () => (
    <ScrollBar progress={0.5} scale={10} style={{ height: '100vh' }} />
  ))
  .add('with progress 0.75', () => (
    <ScrollBar progress={0.75} scale={10} style={{ height: '100vh' }} />
  ))
  .add('with progress 1', () => <ScrollBar progress={1} scale={10} style={{ height: '100vh' }} />);
