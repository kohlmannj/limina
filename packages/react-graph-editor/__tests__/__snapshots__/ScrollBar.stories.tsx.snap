// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-graph-editor/ScrollBar with default props 1`] = `
<Track
  orientation="vertical"
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    scale={1}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar with knobs 1`] = `
<ScrollBar
  orientation="vertical"
  overflow="auto"
  progress={0}
  scale={2}
  style={
    Object {
      "height": "400px",
      "margin": "25vh",
    }
  }
/>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 2 with progress 0 1`] = `
<Track
  orientation="vertical"
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    scale={2}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 2 with progress 0.5 1`] = `
<Track
  orientation="vertical"
  progress={0.5}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={0.5}
    scale={2}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 2 with progress 0.25 1`] = `
<Track
  orientation="vertical"
  progress={0.25}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={0.25}
    scale={2}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 2 with progress 0.75 1`] = `
<Track
  orientation="vertical"
  progress={0.75}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={0.75}
    scale={2}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 2 with progress 1 1`] = `
<Track
  orientation="vertical"
  progress={1}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={1}
    scale={2}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 10 with progress 0 1`] = `
<Track
  orientation="vertical"
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    scale={10}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 10 with progress 0.5 1`] = `
<Track
  orientation="vertical"
  progress={0.5}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={0.5}
    scale={10}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 10 with progress 0.25 1`] = `
<Track
  orientation="vertical"
  progress={0.25}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={0.25}
    scale={10}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 10 with progress 0.75 1`] = `
<Track
  orientation="vertical"
  progress={0.75}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={0.75}
    scale={10}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;

exports[`react-graph-editor/ScrollBar/Vertical/Scale 10 with progress 1 1`] = `
<Track
  orientation="vertical"
  progress={1}
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <ThumbContainer
    orientation="vertical"
    progress={1}
    scale={10}
  >
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="start"
    />
    <ThumbSegment
      orientation="vertical"
      position="middle"
    />
    <ThumbSegment
      dragSignifier={true}
      orientation="vertical"
      position="end"
    />
  </ThumbContainer>
</Track>
`;
